pkg/pub_package_reader/lib/src/archive_surface.dart

```diff
@@ -0,0 +1,51 @@
+// Copyright (c) 2021, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'dart:io';
+
+import 'package:meta/meta.dart';
+import 'package:pub_package_reader/pub_package_reader.dart' show ArchiveIssue;
+
+/// Scans the archive.tar.gz before parsing the tar content:
+/// - for zero length content
+/// - for bad content (parse gzip)
+/// - too large content (gzip or uncompressed)
+Stream<ArchiveIssue> scanArchiveSurface(
+  String archivePath, {
+  @required int maxArchiveSize,
+}) async* {
+  // Some platforms may not be able to create an archive, only an empty file.
+  final file = File(archivePath);
+  if (!file.existsSync() || file.lengthSync() == 0) {
+    yield ArchiveIssue('Package archive is empty (size = 0).');
+    return;
+  }
+
+  // compressed file size check
+  if (file.lengthSync() > maxArchiveSize) {
+    yield ArchiveIssue(
+        'Package archive is too large (size > $maxArchiveSize).');
+    return;
+  }
+
+  int uncompressedLength = 0;
+  try {
+    uncompressedLength = await file
+        .openRead()
+        .transform(gzip.decoder)
+        .fold<int>(0, (length, element) => length + element.length);
```

Comment:
A safer alternative is to use [`limitChunkedStream`](https://pub.dev/documentation/chunked_stream/latest/chunked_stream/limitChunkedStream.html) with a `maxSize` and then handle the `MaximumSizeExceeded` exception.

Because a tiny gzip stream can contain many GB, and this still causes the full stream to be inflated. While it shouldn't eat all memory, it could still take up a lot of CPU.. and effectively take a very long time..

Reply:
Ack, doing it in a follow-up PR.